File: 025_DigitProductEquation.rb
# 025 - Digit Product Equation（★7）
# 再帰関数

N, B = gets.split.map(&:to_i)

@fms = [0]
# fmの候補を作成. mが234も243もf(m)は同じ値となるため探索を省略
def dfs(size, num, nex)
  return @fms << num if size == 0
  nex.upto(9) { dfs(size - 1, num * _1, _1) }
end
dfs(N.to_s.size, 1, 1)

ans = @fms.uniq.count do |fm|
  m = B + fm
  next if m > N
  # fmとfm候補のiが同じになるかチェック
  fm == m.to_s.chars.map(&:to_i).reduce(:*)
end

puts ans

File: kotlin/025_DigitProductEquation.kt
/**
 * 025 - Digit Product Equation（★7）
 * 再帰関数
 */

fun solve() {
  fun f(x: Long) = x.toString().toList().map(Character::getNumericValue).fold(1L) {acc, i -> acc * i}
  val (N, B) = readLine()!!.split(" ").map(String::toLong)

  val fm = mutableListOf<Long>()
  fm.add(0)
  fun dfs(remain: Int, next: Int,  num: Long) {
    if (remain == 0) {
      fm.add(num)
      return
    }
    (next .. 9).forEach { dfs(remain - 1, it, num * it) }
  }
  dfs(N.toString().length, 1, 1L)

  val ans = fm.distinct().count {
    val m = B + it
    if (m <= N) it == f(m) else false
  }

  println(ans)
}

fun main() {
  Thread(null, ::solve, "solve", 1.shl(26)).start()
}
